vllm/model_executor/models/registry.py

```diff
@@ -415,10 +420,94 @@ class _LazyRegisteredModel(_BaseRegisteredModel):
     module_name: str
     class_name: str
 
-    # Performed in another process to avoid initializing CUDA
+    @staticmethod
+    def _get_cache_dir() -> Path:
+        return Path(envs.VLLM_CACHE_ROOT) / "modelinfos"
+
+    def _get_cache_filename(self) -> str:
+        cls_name = f"{self.module_name}-{self.class_name}".replace(".", "-")
+        return f"{cls_name}.json"
+
+    def _load_modelinfo_from_cache(self,
+                                   module_hash: str) -> _ModelInfo | None:
+        try:
+            modelinfo_path = self._get_cache_dir() / self._get_cache_filename()
+            if not modelinfo_path.exists():
+                logger.debug(("Cached model info file "
+                              "for class %s.%s not found"), self.module_name,
+                             self.class_name)
+                return None
+
+            with open(modelinfo_path, encoding="utf-8") as file:
+                mi_dict = json.load(file)
+
+            if mi_dict["hash"] != module_hash:
+                logger.debug(("Cached model info file "
+                              "for class %s.%s is stale"), self.module_name,
+                             self.class_name)
+                return None
+
+            # file not changed, use cached _ModelInfo properties
+            return _ModelInfo(**mi_dict["modelinfo"])
+        except Exception:
+            logger.exception(("Cached model info "
+                              "for class %s.%s error. "), self.module_name,
+                             self.class_name)
+            return None
+
+    def _save_modelinfo_to_cache(self, mi: _ModelInfo,
+                                 module_hash: str) -> None:
+        """save dictionary json file to cache"""
+        try:
+            mi_dict = {}
+            for f in fields(mi):
+                mi_dict[f.name] = getattr(mi, f.name)
+            modelinfo_dict = {
+                "hash": module_hash,
+                "modelinfo": mi_dict,
+            }
+            modelinfo_path = self._get_cache_dir() / self._get_cache_filename()
```

Comment:
If we do the rename, we just have to save to a new tmp file and then rename it to `modelinfo_path`

Reply:
Implemented an `atomic_writer` utility function